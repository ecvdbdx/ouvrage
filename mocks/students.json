[
  {
    "username": "epatrois",
    "first_name": "Manu",
    "last_name": "Patrois",
    "id": "2"
  },
  {
    "id": "5",
    "username": "mdewaele",
    "first_name": "Maxence",
    "last_name": "De Waele",
    "email": "maxencedewaele@gmail.com",
    "birthdate": "838921463",
    "avatar": "https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png"

  },
  {
    "id": "8",
    "username": "rodrigodelanoche",
    "first_name": "Vincent",
    "last_name": "Rodriguez",
    "email": "vincentrodrigu@hotmail.fr",
    "birthdate": "14/05/1413",
    "avatar": "http://etc.usf.edu/clipart/5500/5596/columbus_14_md.gif"
  },
  {
    "id": "6",
    "username": "erozis",
    "first_name": "Emilie",
    "last_name": "Rozis",
    "email": "emilie.40@hotmail.com"
  },
  {
    "first_name": "Hubert",
    "username": "hmoncenis",
    "id": "1",
    "last_name": "Montcenis"
  },
  {
    "id": "7",
    "username": "vivianeglz",
    "first_name": "Viviane",
    "last_name": "Gonzalez",
    "email": "viviane.gonzalez@outlook.fr",
    "birthdate": "02/06/1995",
    "avatar": "kikou.png"
  },
  {
    "id": "3",
    "username": "antoinesinety",
    "first_name": "Antoine",
    "last_name": "de Sinety",
    "email": "fake@yopmail.fr",
    "birthdate": "08/06/1995",
    "avatar": "Image.png"
  },
  {
    "id": "4",
    "username": "tgauzin",
    "first_name": "Thomas",
    "last_name": "Gauzin",
    "email": "thomas.gauzin@gmail.com"
  }
]
