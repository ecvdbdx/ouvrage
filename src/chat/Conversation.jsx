import React from 'react';

export default class Conversation extends React.Component {
  render () {
    const currentContact = this.props.currentContact;
    return (
      <div id="conversation">
        {
          currentContact ?
            <div className={currentContact === null ? 'hide' : ''}>
              <header>
                <h2>
                  { currentContact.username }
                </h2>
              </header>
              <section>
                <h3>
                  <div className="user-avatar" >
                    <img src={currentContact.avatar}/>
                  </div>
                  <div className="user-info">
                    <a>{currentContact.username}</a>
                  </div>
                </h3>
              </section>
            </div> : <div/>
        }
      </div>
    );
  }
}
