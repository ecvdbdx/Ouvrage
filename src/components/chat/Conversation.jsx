import React from 'react';
import Message from './Message.jsx';
import ChatApi from './ChatApi';

export default class Conversation extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      message: '',
      messages: [],
      autoScrollActive: false
    };
    this.submitMessage = this.submitMessage.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.scrollDown = this.scrollDown.bind(this);
    this.setAutoScroll = this.setAutoScroll.bind(this);

    this.refMessageList = React.createRef();

    if (this.props.currentContact) {
      this.retrieveConversation(this.props.currentContact, this.props.me);
      setInterval(() => {
        this.retrieveConversation(this.props.currentContact, this.props.me);
      }, 2000);
    }
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.currentContact !== this.props.currentContact)
      this.retrieveConversation(nextProps.currentContact, this.props.me);
  }

  retrieveConversation (contact, me) {
    ChatApi.getConversation(contact, me).then((messages) => {
      this.setState({
        messages: messages
      });
      this.scrollDown();
    });
  }

  handleChange (event) {
    this.setState({
      message: event.target.value
    });
  }


  submitMessage (event) {
    if (event.key === 'Enter') {
      const textField = event.target;
      const message = textField.value;
      const me = this.props.me;
      const user = this.props.currentContact;
      ChatApi.sendMessage(message, me, user).then((record) => {
        let messages = this.state.messages;
        messages.push(record.fields);
        this.setState({
          messages: messages,
          message: ''
        });
        textField.value = '';
        this.scrollDown();
      });
    }
  }

  scrollDown () {
    if (this.state.autoScrollActive) {
      const nodeElement = this.refMessageList.current;
      nodeElement.scrollTop = nodeElement.scrollHeight;
    }
  }

  setAutoScroll (event) {
    const nodeElement = this.refMessageList.current;
    if (nodeElement === null) return 0;
    const actuallScroll = nodeElement.scrollHeight - nodeElement.offsetHeight;
    if (actuallScroll - nodeElement.scrollTop <= 3) {
      this.setState({
        autoScrollActive: true
      });
    }
    else {
      this.setState({
        autoScrollActive: false
      });
    }
  }

  render () {
    const currentContact = this.props.currentContact;
    const me = this.props.me;
    if (!currentContact) return null;

    return (
      <section className="conversation" ref={this.refMessageList} onScroll={this.setAutoScroll}>
        {
          <div>
            <header>
              <h2>{currentContact.firstName} {currentContact.lastName}</h2>
              <img src={currentContact.avatar} className="user-avatar"/>
            </header>
            <main>
              <ul className="message-list">
                {
                  this.state.messages.map((message, index) => (
                    <Message key={index} message={message} me={me}/>
                  ))
                }
              </ul>
            </main>
            <footer className="message-section">
              <input type="text"
                placeholder="Votre message..."
                onChange={this.handleChange}
                value={this.state.message}
                onKeyPress={this.submitMessage}
              />
            </footer>
          </div>
        }
      </section>
    );
  }
}
