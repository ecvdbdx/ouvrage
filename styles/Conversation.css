#conversation header{
    border-bottom: 1px solid rgba(0, 0, 0, .10);
    box-sizing: border-box;
    height: 50px;
    padding: 8px;
    position: relative;
    text-align: center;
    margin-left: -12px;
    margin-right: -12px;
}

#conversation {
    border-bottom: 1px solid rgba(0, 0, 0, .05);
    min-height: 50px;
    padding: 12px;
    position: relative;
}

#conversation .user-avatar{
    align-items: center;
    background-color: rgba(0, 0, 0, .05);
    border-radius: 50%;
    display: flex;
    justify-content: center;
    overflow: hidden;
    height: 60px;
    width: 60px;
    display: inline-block;
}

#conversation .user-avatar img{
    height: 100%;
    width: 100%;
}
#conversation .user-info{
    display: inline-block;
    vertical-align: top;
    padding-left: 10px;
}
#conversation h3{
    display: inline-block;
    vertical-align: top;
    padding-left: 10px;
    margin-top: 5px;
}
#conversation .user-info a{
    color: #365899;
    cursor: pointer;
    text-decoration: none;
    font-size: 16px;
}