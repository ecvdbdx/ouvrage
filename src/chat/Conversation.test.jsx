import React from 'react';
import renderer from 'react-test-renderer';
import students from './../../mocks/students.json';
import Conversation from './Conversation.jsx';

it('renders correctly', () => {
  const tree = renderer
    .create(<Conversation currentContact={students[0]}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
