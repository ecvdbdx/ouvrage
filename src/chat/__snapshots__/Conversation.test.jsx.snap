// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="conversation"
>
  <h2>
    La conversation
  </h2>
  epatrois
</div>
`;
